File: LeetCode/CPP/Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        /*
        for(int i = 0; i < s.size(); i++) {
            if(s[i] == '(' || s[i] == '{' || s[i] == '[') {
                st.push(s[i]);
            } else {
                if(st.empty()) {
                    return false;
                }
                if(s[i] == ')' && st.top() == '(') {
                    st.pop();
                } else if(s[i] == '}' && st.top() == '{') {
                    st.pop();
                } else if(s[i] == ']' && st.top() == '[') {
                    st.pop();
                } else {
                    return false;
                }
            }
        }
        return st.empty();
        */
        for(int i = 0; i < s.size(); i++) {
            if(s[i] == '(') {
                st.push(')');
            } else if(s[i] == '{') {
                st.push('}');
            } else if(s[i] == '[') {
                st.push(']');
            } else {
                if(st.size() && st.top() == s[i]) {
                    st.pop();
                } else {
                    return false;
                }
            }
        }
        return st.empty();
    }
};

File: LeetCode/JS/Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let temp = [];
    if(s.length %2) return false;

    for(let i = 0; i< s.length; i++) {
        if(s[i] === "(" || s[i] === "{" || s[i] === "[") {
            temp.push(s[i]);
        } else {
            const lastItem = temp.pop();
            if(lastItem === '(' && s[i] === ')') continue;
            if(lastItem === '[' && s[i] === ']') continue;
            if(lastItem === '{' && s[i] === '}') continue;

            return false;
        }
    }

    return temp.length === 0;
};

//Clean solution
var isValid = function(s) {
    let temp = [];
    if(s.length %2) return false;

    for(let i = 0; i< s.length; i++) {
        if(s[i] === "(") {
            temp.push(")");
        } else if(s[i] === "{") {
            temp.push("}");
        } else if(s[i] === "[") {
            temp.push("]");
        } else {
            if(temp.pop() !== s[i]) return false;
        }
    }

    return temp.length === 0;
};
